sklearn/metrics/pairwise.py

```diff
@@ -687,7 +613,245 @@ def cosine_distances(X, Y=None):
+        is assumed the data is already scaled.
+
+    **metric_kwargs : dict, optional
+        Keyword arguments with precomputed parameters (MIN, MAX)
+        to scale the data.
+        
+    Returns
+    -------
+    similarities : ndarray, shape (n_samples, n_samples)
+
+    References
+    ----------
+    Gower, J.C., 1971, A General Coefficient of Similarity and Some of Its
+    Properties.
+
+    Notes
+    -----
+    The numeric feature ranges are determined from both X and Y.
+
+    Current implementation does not support sparse matrices.
+
+    All the non-numerical types (e.g., str), are treated as categorical
+    features.
+
+    This implementation modifies the Gower's original similarity measure in
+    the folowing aspects:
+
+    * The values in the original similarity S range between 0 and 1. To
+    guarantee this, it is assumed the numerical features of X and Y are
+    scaled between 0 and 1.
+
+    * Different from the original similarity S, this implementation
+    returns 1-S.
+
+    """
+    if issparse(X) or issparse(Y):
+        raise TypeError("Gower distance does not support sparse matrices")
+
+    if X is None or len(X) == 0:
+        raise ValueError("X can not be None or empty")
+
+    y_is_not_none = Y is not None
+
+    # It is necessary to convert to ndarray in advance to define the dtype
+    # as np.object, otherwise numeric columns will be converted to string
+    # if there are other string columns.
+    if not isinstance(X, np.ndarray):
+        X = np.asarray(X, dtype=np.object)
+
+    if y_is_not_none and not isinstance(Y, np.ndarray):
+        Y = np.asarray(Y, dtype=np.object)
+
+    X, Y = check_pairwise_arrays(X, Y, precomputed=False, dtype=X.dtype,
+                                 force_all_finite=False)
+
+    n_rows, n_cols = X.shape
+
+    categorical_features_obj, categorical_features_num, numerical_features = \
+        _detect_categorical_features(X, categorical_features)
+
+    # Categorical columns
+    X_cat_obj = X[:, categorical_features_obj].astype(np.object)
+    X_cat_num = X[:, categorical_features_num].astype(np.float)
+
+    # Numerical columns
+    X_num = X[:, numerical_features].astype(np.float)
+
+    Y_cat_obj = None
+    Y_cat_num = None
+    Y_num = None
+    if y_is_not_none:
+        Y_cat_obj = Y[:, categorical_features_obj].astype(np.object)
+        Y_cat_num = Y[:, categorical_features_num].astype(np.float)
+        Y_num = Y[:, numerical_features].astype(np.float)
+    else:
+        Y_cat_obj = X_cat_obj
+        Y_cat_num = X_cat_num
+        Y_num = X_num
+
+    num_cols = X_num.shape[1]
+    n_col_num_present = num_cols > 0
+    n_col_cat_obj_present = X_cat_obj.shape[1] > 0
+    n_col_cat_num_present = X_cat_num.shape[1] > 0
+
+    # Calculates the min and max values, and if requested, scale the
+    # input values as proposed by the Gower's paper.
+    if n_col_num_present:
+        MIN_MAX = {}
```

Comment:
I don't think we *need* both min and max. I can't think of any reason (except perhaps for clipping new data) to require variables scaled between 0 and 1, as long as their total range is 1 (i.e. `np.all(ptp(X_scaled, axis=0) == 1)`).

Reply:
Makes sense, I'll use only the range, and see if the tests brake.

Reply:
We are loosing information when using only the range of numerical variables, all the numerical tests are breaking. 

I can't fit the MinMaxScaler with ranges only (we already agreed MinMaxScaler will be the best option because it is able to deal with nan values). If I use the X, Y fit there is no space for ranges, if I use the min of X, Y,  together with min of X, Y + ranges maybe the min of X,Y is not the min that generated initially the ranges.

Checking inside MinMaxScaler fit method I can see it looks for min and max values, the ranges won't bring this information. We need to have at least min values, better to have also the max values as parameters for scaling, instead of ranges.

Ill see what I can do with min and max here, please let me know if someone has a better idea.

Reply:
Sorry, I was wrong! 

We can surely use only the ranges to scale, once we internally take a min value from the data we have now, no matter this data is a subset from the original data, as long the ranges represent the original data. I have tests supporting this, pushing the solution with ranges today.